### PR TITLE
CoqIDE: Improve preferences dialog

### PR DESCRIPTION
**Pull request by Frigory33**
Hello. This is quite a huge pull request, which half rewrites the code for the preferences dialog. The main problem was that the code was apparently based on a copy of a small library named Configwin. This made the preferences dialog more complicated to improve: the “Configwin” wasn’t configurable enough and offered many useless things. It wasn’t very adapted. So I wrote a more concise module which has what the preference dialog actually uses.

The preference dialog then has some style and feature improvements, as explained in the second commit’s message.

To update pictures in the documentation, I took screenshots showing my GTK theme. I hope that’s not a problem.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.

### CODE REVIEW

File: doc/sphinx/_static/coqide-preferences-editor.png
Comment (jfehrle) [inline doc/sphinx/_static/coqide-preferences-editor.png:1]: Would you redo the screen shot using light theme to be consistent with the other screen shots of CoqIDE?  Also the button names at the bottom should be in English.  While I expect you prefer the dark theme, consistency is more important.  I recommend not trying to change everything to dark theme, either--sometimes we need to set aside our individual preferences.

File: doc/sphinx/_static/coqide-preferences-shortcuts.png
Comment (jfehrle) [inline doc/sphinx/_static/coqide-preferences-shortcuts.png:1]: Also please make this one light theme

**Review by jfehrle — COMMENTED** [review]
I took a very quick look and made a few comments.  Generally IMO reviewing IDE changes must include testing--and often on multiple platforms.  IIRC you develop on Mac.  We should test this both on Linux and Windows, which I can do but will take a while.  As well as examining the code changes.

File: doc/changelog/10-coqide/19417-coqide-preferences.rst [hunk 1]
```diff
@@ -0,0 +1,20 @@
+- **Changed:**
+  Preferences dialog is improved: more margins,
+  tree of categories, sections in the categories,
+  spin buttons for numbers, preservation of the last
+  selected category, and more
+  (`#19417 <https://github.com/coq/coq/pull/19417>`_,
+  by Sylvain Chiron).
+- **Fixed:**
+  No more preferences are applied before clicking Apply or OK
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:9]: I’m not sure about this: “Apply” was useful for most preferences. Only a few were applied immediately. Well, let’s say “All preferences” then.

File: doc/sphinx/_static/coqide.png
[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/sphinx/_static/coqide.png:1]: Well, I thought it would be great to show that CoqIDE can be used with a dark theme.

The old screenshot has some obsolete components: a “Try Tactics” menu, “.v” in the name of the buffer, and tool bar buttons are not organized the same way.

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/sphinx/_static/coqide.png:1]: By the way, could we find some place in the repository where we could put example files, so that we don’t have to type them again when we want to update the screenshots?

[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/sphinx/_static/coqide.png:1]: I already put a couple of them as comments within the rst file--less likely to get separated.  Maybe somewhere under `test-suite` even though they're not examples.  And include the path to the file in the rst.

File: doc/changelog/10-coqide/19417-coqide-preferences.rst [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+- **Fixed:**
+  Changing the allowed modifiers in the Shortcuts panel of
+  the Preferences dialog now immediately updates the available
+  modifiers for the listed menus
+  (`#19417 <https://github.com/coq/coq/pull/19417>`_,
+  by Sylvain Chiron).
+- **Added:**
+  Preference setting for unjustified conclusions background color
+  (`#19417 <https://github.com/coq/coq/pull/19417>`_,
+  by Sylvain Chiron).
```
Comment (jfehrle) [inline doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: On the shortcuts menu, if the user clicks on `Alt` in the allowed modifiers by mistake and then clicks again on it to restore it, `Alt` is removed from all the entries.  It would be friendlier to remember/reshow the previous settings as the user may not notice that they were reset.  If that's not too hard to do.

Comment (jfehrle) [inline doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: It seems odd to have both an Appearance tab and "Editor appearance" on the Editor tab.  Maybe the editor ones could move to the Appearance tab?  Maybe move the window side ones to the bottom of the panel?

**Review by jfehrle — COMMENTED** [review]
Comments on the appearance/behavior of the dialog.  It does look considerably better.

Some of my comments suggest things to definitely try to fix.  Others are probably beyond what you want to take on in this PR.  Let me know your thoughts and about anything you disagree on.

I haven't reviewed the code yet.

File: doc/changelog/10-coqide/19417-coqide-preferences.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: > This looks odd. It didn't say that before. Previously it said `coq_style`. Would you change it to a better name?

It is in the `coq_style.xml` file. So let’s say “Coq style”?

> I'm also getting odd behavior when I click repeatedly on Style scheme. Frequently, when the dropdown isn't showing, the dropdown appears when I click and hold the mouse button and disappears when I release the button. Do you see the same behavior? It may be an artifact of my X-Windows, which is not very good.

I don’t get any strange behavior; the dropdown list appears all the time for me. If you hold the mouse button, then move the mouse and release the button, it is normal that the dropdown disappears immediately.

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: I don’t know, I don’t have this bug! It is not a new bug, isn’t it?

This is a full-featured control of GTK so we cannot configure it much.

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: I don’t know. “Editor appearance” seems to decide how the editor shows text and other things, while “Editor behavior” is what happens when you type. According to this, “Dynamic word wrap” is an appearance setting.

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: It looks hard to fix this. The tree indents only elements of its first column, and there is a column for icons and another for text.

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: No, I don’t see this, I simply cannot move the dialog box while a dropdown is open.

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: > On the Files menu, arguably the Global Auto Reload and Auto Save options should go on the Editor panel. And probably put Auto Save first, IMO that's the more common operation.

Well, I think it’s nice as it is.

> FYI: if you click in the menu and type characters, you get a box like this at the bottom. I expected it was a search feature and the pressing Enter would make it jump to an item with that name. But it doesn't jump anywhere AFAIK. Not new in your code.

Yep, that’s weird, it seems it doesn’t do anything. It’s a feature of GTK. Probably it should select the corresponding item of the tree according to the first column, but the first column contains only icons.

[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: I suppose.  FWIW if you don't show line numbers or highlight the current line, it's hard to distinguish where lines really begin (ie \n).

[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: How about this?

[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: I don't see it today.  Probably it's an artifact of my crummy X-windows software.

[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: I suppose good enough.  I think the Language setting is supposed to choose between `ide/coqide/coq.lang` and `.../coq-ssreflect`, which specifies how to do syntax highlighting.  I thought `using` would be highlighted as an SSR keyword, but I can't get it to happen. I didn't try very hard and these files look complex.  And I'm sure they're out of date (missing commands, tactics and keywords) not to mention that regular expressions can't do a perfect job highlighting keywords.

image

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: Yes, we already discussed last year that they are out of date, and it was also said that regular expressions cannot do a perfect job — though I still wonder why regexp aren’t enough, and PCREs allow a bit more than regexps, there are some look-ahead syntaxes allowed.

[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: Okay, I’ve applied this idea.

[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: Hmm, maybe too strong to say regexps can't do a perfect job.  I don't have a good example in hand.  Depends on what you want to highlight.

Adding lookahead to REs doesn't make them more powerful in theory.  There's always an equivalent non-lookahead RE.  But REs with lookahead may make them easier to work with.  REs are not able to correctly recognize only balanced parentheses, e.g. accepting "(())()" but not "(()".  To do that requires a context-free grammar and a parser stack.

File: doc/sphinx/proof-engine/ltac2.rst [hunk 3]
```diff
@@ -860,6 +860,8 @@ in a less hard-wired way.
       dependencies in the variables :n:`@ident__i` and returns a functional term
       of the form :n:`fun @ident__1 … @ident__n => @term`.
 
+   .. _match_term_context_2:
+
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/ltac2.rst:864]: Why is this needed?  I didn't find any link referring to it.

File: ide/coqide/preferences_ui.ml [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+
+  let misc =
+    create_pref_box "Miscellaneous" [
+        pbool "Stop interpreting before the current point" stop_before;
+        pbool "Reset coqtop on tab switch" reset_on_tab_switch;
+      ]
+  in
+
+(*
+  let add_user_query () =
```
Comment (jfehrle) [inline ide/coqide/preferences_ui.ml:643]: Is there a reason to keep this commented-out code? If so, maybe add a short comment explaining why.

File: ide/coqide/coqide.ml [hunk 5]
```diff
@@ -340,10 +340,10 @@ let check_quit ?parent saveall =
...
   let is_modified sn = sn.buffer#modified in
   if List.exists is_modified notebook#pages then begin
-    let answ = Configwin_ihm.question_box ~title:"Quit"
-      ~buttons:["Save Named Buffers and Quit";
-                "Quit without Saving";
-                "Don't Quit"]
+    let answ = Ui_dialogs.question_box ~title:"Quit"
+      ~buttons:[ButtonWithLabel "Save named buffers and quit";
+                ButtonWithLabel "Quit without saving";
+                ButtonWithStock `CANCEL]
```
Comment (jfehrle) [inline ide/coqide/coqide.ml:346]:
```suggestion
                ButtonWithLabel "Don't Quit"]
```

I think the original button text is clearer.

File: ide/coqide/preferences_ui.ml [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+    create_pref_box ~in_grid:true "Window" [
+      pint "Width at startup:" window_width;
+      pint "Height at startup:" window_height;
+    ]
+  in
+  let config_document_tabs =
+    create_pref_box "Document tabs" [
+        let positions = ["top"; "left"; "right"; "bottom"] in
+        let labels = ["Top"; "Left"; "Right"; "Bottom"] in
+        let initial = List.nth labels (match find_index_opt document_tabs_pos#get positions with Some i -> i | _ -> 0) in
```
Comment (jfehrle) [inline ide/coqide/preferences_ui.ml:454]:
```suggestion
        let initial = List.nth labels
            (match find_index_opt document_tabs_pos#get positions with Some i -> i | _ -> 0) in
```

File: doc/sphinx/practical-tools/coqide.rst [hunk 7]
```diff
@@ -23,21 +23,39 @@ for CoqIDE.  Use `coqide --help` to see the list of command line options.
...
 
-The lower right panel has three tabs: the *Messages
-panel*, which shows messages produced by commands and tactics; the *Errors panel*,
-which shows errors detected when running in :ref:`async mode <asyncmode>` and the
-*Jobs panel,* which shows information on the worker processes used by async mode.
-The contents of the right-hand panels are specific to the currently-displayed script.
+The lower right panel has three tabs:
+
+- the *Messages panel* shows messages produced by commands and tactics;
+- the *Errors panel* shows errors detected when running in :ref:`async mode <asyncmode>`;
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:53]: Could you mention the Warnings tab, too?

File: doc/sphinx/practical-tools/coqide.rst [hunk 8]
```diff
@@ -23,21 +23,39 @@ for CoqIDE.  Use `coqide --help` to see the list of command line options.
...
+- the *Jobs panel* shows information on the worker processes used by async mode.
+
+The contents of the right-hand panels are specific to the currently displayed script.
 Click the arrow icons to detach these panel into separate windows.  The proof panel
-can be detached from the `Windows/Detach Proof` menu item.
+can be detached from the *Windows/Detach Proof* menu item.
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:58]:
```suggestion
can be detached with the *Windows/Detach Proof* menu item.
```

File: doc/sphinx/practical-tools/coqide.rst [hunk 9]
```diff
@@ -209,51 +228,48 @@ the *View* menu, or using the shortcut ``F2``.
...
 Compilation
 -----------
 
-The `Compile` menu offers direct commands to:
+The *Compile* menu offers direct commands to:
 
-+ compile the current buffer
-+ run a compilation using `make`
-+ go to the next compilation error and
++ compile the current buffer;
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:239]: FYI Semicolons or commas between bullet items is optional in English except for formal writing such as academic papers or government documents.

File: doc/sphinx/practical-tools/coqide.rst [hunk 10]
```diff
@@ -265,7 +281,30 @@ Preferences
...
+The *Project* section enables you to change the default name for
+project files and the way that project file options are used.
+
+The *Editor* section (shown in the screenshot above) is for
+customizing the editor. It includes in particular the ability
+to activate an Emacs mode named micro-Proof-General
+(use the Help menu to know more about the available bindings).
+
+The *Appearance* section offers controls to set CoqIDE's window
+default size and the position of tabs.
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:307]: Do you want to briefly cover all the sections?  You'd need to add Font, Colors, Tags, Shortcuts and Misc.  OTOH, I think most users are clever enough to discover these on their own.  Maybe just point out any "unusual" items.

File: doc/sphinx/practical-tools/coqide.rst [hunk 11]
```diff
@@ -321,34 +342,43 @@ The top of the *Shortcuts* section lets you select the allowed modifiers
 that can be selected for the listed menus.  (The changes won't appear until
 you close and reopen the Preferences dialog.)
 
-*Misc*
+*Misc* – to be documented
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:345]:
```suggestion
*Misc*
```

"to be documented" not really needed; no one will be confused.

File: doc/sphinx/practical-tools/coqide.rst [hunk 12]
```diff
@@ -321,34 +342,43 @@ The top of the *Shortcuts* section lets you select the allowed modifiers
...
+the associated item from the menu.  On some systems, you can modify the
+key binding ("accelerator") for a menu entry by going to the corresponding
+menu item without releasing the mouse button, pressing the keys you want
+for the new binding and then releasing the mouse button.
 
-Alternatively, you can edit the file directly.  Make sure there are no
-CoqIDE processes running while you edit the file.  (CoqIDE creates or
-overwrites the file when it terminates, which may reorder the lines).
+Alternatively, you can edit the configuration file directly.
+A GTK+ accelerator keymap is saved under the name `coqide.keys` in
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:377]:
```suggestion
Key bindings are saved in the file `coqide.keys` in
```

This file was just mentioned 25 lines above...

File: doc/sphinx/practical-tools/coqide.rst [hunk 13]
```diff
@@ -369,16 +399,16 @@ and set the third item in the list as desired, such as in the third line.
 Avoid assigning the same binding to multiple items.
 
 If the same menu item name appears on multiple lines in the file, the value from the
-last line is used.  This is convenient for copying a group of changes from elsewhere--just
-insert the changes at the end of the file.  The next time CoqIDE terminates, it will
+last line is used.  This is convenient for copying a group of changes from elsewhere –
+just insert the changes at the end of the file.  The next time CoqIDE terminates, it will
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:403]:
```suggestion
last line is used.  This is convenient for copying a group of changes from elsewhere–just
insert the changes at the end of the file.  The next time CoqIDE terminates, it will
```

The doc follows American usage (not my choice BTW), which doesn't use spaces before and after an en dash.  While I couldn't find an internet reference, you can see this usage in the Chicago Manual of Style.

File: doc/sphinx/practical-tools/coqide.rst [hunk 14]
```diff
@@ -594,21 +624,21 @@ Toggle breakpoint (F8)
...
+  equivalent to "Run to end" (Alt-End).
 
-Step over (Control ↓)
+Step over (Alt-↓)
   When stopped in the debugger,
   execute the next tactic without stopping inside it.  If the debugger reaches
   a breakpoint in the tactic, it will stop.  This is the same key combination used
-  for "Forward one command"—if you're stopped in the debugger then it does a "Step over"
-  and otherwise it does a "Forward".  Combining the two functions makes it easy
+  for *Forward one command* — if you're stopped in the debugger then it does a *Step over*
```
Comment (jfehrle) [inline doc/sphinx/practical-tools/coqide.rst:633]:
```suggestion
  for *Forward one command*—if you're stopped in the debugger then it does a *Step over*
```

**Review by jfehrle — COMMENTED** [review]
The code looks good.  Many tiny suggestions.

I'll plan to merge later this week after you've dealt with the suggestions.

**Comment by jfehrle** [comment]
One other thing: we use several terms for "key bindings", including "Shortcuts" and "menu item accelerators".  Lets at least drop "accelerator", which is used in only 2 places in the doc.

image

File: doc/sphinx/proof-engine/ltac2.rst (hunk 3 above)
Comment (Frigory33) [reply doc/sphinx/proof-engine/ltac2.rst:864]: I think this is the rest of a change I made to avoid a compilation warning (treated as a fatal error). It is apparently useless now, so I’m removing this.

File: ide/coqide/preferences_ui.ml (hunk 4 above)
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml:643]: It’s not from me, it comes from preferences.ml, so I don’t know.

File: ide/coqide/preferences_ui.ml [hunk 15]
```diff
@@ -0,0 +1,690 @@
...
+
+  let config_font =
+    let preview_text = "Goal (∃n : nat, n ≤ 0)∧(∀x,y,z, x∈y⋃z↔x∈y∨x∈z)." in
+    select_arch (
+      let box = GPack.hbox () in
+      let font_sel = GMisc.font_selection ~preview_text:preview_text () in
+      let realized = ref false in (* The font selector does not allow setting its font before it is realized (bug) *)
+      box#pack ~expand:true font_sel#coerce;
+      let _ = font_sel#misc#connect#realize
+                ~callback:(fun () -> font_sel#set_font_name text_font#get; realized := true;) in
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml:468]: Is it related to this part of the code? I don’t understand.

File: ide/coqide/preferences.ml [hunk 16]
```diff
@@ -81,20 +82,18 @@ let stick (pref : 'a preference) (obj : < connect : #GObj.widget_signals ; .. >)
...
-    | `MOD4 -> "<Mod4>"
-    | `MOD5 -> "<Mod5>"
-    | `CONTROL -> "<Control>"
-    | `SHIFT -> "<Shift>"
-    | `HYPER -> "<Hyper>"
-    | `META -> "<Meta>"
-    | `RELEASE -> ""
-    | `SUPER -> "<Super>"
-    |  `BUTTON1| `BUTTON2| `BUTTON3| `BUTTON4| `BUTTON5| `LOCK -> ""
+let mod_to_str = function
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences.ml:85]: Well, it’s better if we use `function`, isn’t it?

File: ide/coqide/preferences_ui.ml (hunk 15 above)
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml:468]: I don’t get this 3 second delay. But there’s still a 0.5 second delay for loading the Fonts panel the first time it’s shown after opening the preferences.

File: ide/coqide/preferences_ui.ml [hunk 17]
```diff
@@ -0,0 +1,690 @@
...
+      ~eq:(fun (q1, k1) (q2, k2) -> k1 = k2 || q1 = "" || q2 = "" || q1 = q2)
+      ~add:add_user_query
+      ~titles:["User query"; "Shortcut key"]
+      "User queries"
+      (fun (q, s) -> [q; s])
+      user_queries#get
+  in
+*)
+
+  let cmds = [
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml:673]: I remember I fixed this, but apparently I felt it was unnecessary to write a comment explaining how…! It was just a matter of putting something in the right order. Now, the Fonts section is no longer the first one.

File: ide/coqide/preferences_ui.ml [hunk 18]
```diff
@@ -0,0 +1,690 @@
...
+          pref_section "Tags" ~icon:`SELECT_COLOR [config_tags];
+        ];
+      pref_section "Shortcuts" ~icon:`MEDIA_FORWARD [modifiers_valid; config_modifiers];
+      pref_section "Externals" ~icon:`EXECUTE [externals_cmds];
+      pref_section "Misc" ~icon:`PREFERENCES [misc];
+    ]
+  in
+
+  if edit_preferences ~apply "Preferences" ~parent ~current_section ~width:600 ~height:400 cmds then
+    save_pref ();
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml:690]: OK.

File: ide/coqide/ui_dialogs.ml [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        | ButtonWithLabel label -> Some label, None
+        | ButtonWithStock stock -> None, Some stock
+      in
+      let but = GButton.button ?label ?stock ~packing:(bbox#pack ~expand:true) () in
+      ignore (but#connect#clicked ~callback:(fun () -> button_nb := pos; window#destroy ()));
+      (* If it's the default button then give it the focus *)
+      if pos = default then
+        but#grab_default ();
+      iter_buttons (pos + 1) others;
+  | _ -> ()
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/ui_dialogs.ml:36]: OK.

File: ide/coqide/coqide.ml (hunk 5 above)
Comment (Frigory33) [reply ide/coqide/coqide.ml]: OK, restored.

File: ide/coqide/coqide.ml [hunk 20]
```diff
@@ -472,10 +472,10 @@ let close_buffer ?parent sn =
...
   if not sn.buffer#modified then do_remove ()
   else
-    let answ = Configwin_ihm.question_box ~title:"Close"
-      ~buttons:["Save Buffer and Close";
-                "Close without Saving";
-                "Don't Close"]
+    let answ = Ui_dialogs.question_box ~title:"Close"
+      ~buttons:[ButtonWithLabel "Save buffer and close";
+                ButtonWithLabel "Close without saving";
+                ButtonWithStock `CANCEL]
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/coqide.ml]: OK, restored.

File: ide/coqide/preferences_ui.ml [hunk 21]
```diff
@@ -0,0 +1,690 @@
...
+        apply ();
+        false
+      in
+      tree#foreach foreach;
+  }
+
+let edit_preferences
+    ?(apply = fun () -> ())
+    title ?parent ?width ?height ?(current_section = ref None)
+    pref_struct =
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: OK, applied.

File: ide/coqide/preferences_ui.ml [hunk 22]
```diff
@@ -0,0 +1,690 @@
...
+          pref_section "Colors" ~icon:`SELECT_COLOR [config_highlight; config_color];
+          pref_section "Tags" ~icon:`SELECT_COLOR [config_tags];
+        ];
+      pref_section "Shortcuts" ~icon:`MEDIA_FORWARD [modifiers_valid; config_modifiers];
+      pref_section "Externals" ~icon:`EXECUTE [externals_cmds];
+      pref_section "Misc" ~icon:`PREFERENCES [misc];
+    ]
+  in
+
+  if edit_preferences ~apply "Preferences" ~parent ~current_section ~width:600 ~height:400 cmds then
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: Yep. Applied.

File: ide/coqide/preferences_ui.ml [hunk 23]
```diff
@@ -0,0 +1,690 @@
...
+    create_pref_box "Global auto reload" [
+      pbool "Enable" global_auto_reload;
+      pint "Delay (ms):" global_auto_reload_delay;
+    ]
+  in
+  let auto_save =
+    create_pref_box "Auto save" [
+      pbool "Enable" auto_save;
+      pint "Delay (ms):" auto_save_delay;
+      (* auto_save_name *)
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: OK, done.

File: ide/coqide/preferences_ui.ml [hunk 24]
```diff
@@ -0,0 +1,690 @@
...
+        let labels = ["Top"; "Left"; "Right"; "Bottom"] in
+        let initial = List.nth labels (match find_index_opt document_tabs_pos#get positions with Some i -> i | _ -> 0) in
+        let setter i _ = document_tabs_pos#set (List.nth positions i) in
+        pcombo "Position:" labels initial setter;
+      ]
+  in
+
+  let config_font =
+    let preview_text = "Goal (∃n : nat, n ≤ 0)∧(∀x,y,z, x∈y⋃z↔x∈y∨x∈z)." in
+    select_arch (
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: I didn’t write an `is_mac` function, but this is done.

File: ide/coqide/preferences_ui.ml [hunk 25]
```diff
@@ -0,0 +1,690 @@
...
+      pbool "Enable" auto_save;
+      pint "Delay (ms):" auto_save_delay;
+      (* auto_save_name *)
+    ]
+  in
+
+  let project_management =
+    create_pref_box "Project management" [
+        pstring "Default name for project file:" project_file_name;
+        pcombo "Project file options are" (List.map read_project#repr#raw_from [Subst_args; Append_args; Ignore_args])
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: OK, fixed.

File: ide/coqide/preferences_ui.ml [hunk 26]
```diff
@@ -0,0 +1,690 @@
...
+        apply_list := (fun () -> pref#set (Gdk.Color.color_to_string button#color)) :: !apply_list;
+        let _ = reset_button#connect#clicked ~callback:(fun () ->
+            pref#reset ();
+            button#set_color (Gdk.Color.color_parse pref#get);
+          )
+        in
+        ()) prefs;
+    in
+    let () = Util.List.iteri iter [
+      (*"Editor", [background_color]; *)
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: OK, removed.

File: ide/coqide/preferences_ui.ml [hunk 27]
```diff
@@ -0,0 +1,690 @@
...
+    let box = GPack.vbox ~packing:(menu_box#pack ~expand:true) ~show:false () in
+    box#set_margin_left 4;
+    let section_widgets = List.map (make_section box) sections in
+    let box_widget = {
+      widget = box#coerce;
+      apply = fun () -> List.iter (fun { apply } -> apply ()) section_widgets;
+    } in
+    let new_iter = tree#append ?parent () in
+    Option.iter (tree#set ~row:new_iter ~column:icon_col) icon;
+    (* Without this, labels are not indented *)
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: OK, changed.

File: doc/changelog/10-coqide/19417-coqide-preferences.rst [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+- **Fixed:**
+  All preferences are now applied after clicking Apply or OK
+  rather than immediately
+  (`#19417 <https://github.com/coq/coq/pull/19417>`_,
+  by Sylvain Chiron).
+- **Fixed:**
+  Changing the allowed modifiers in the Shortcuts panel of
+  the Preferences dialog now immediately updates the available
+  modifiers for the listed menus
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst]: I don’t understand your suggestion as these are menus. But OK.

File: ide/coqide/preferences_ui.ml (hunk 6 above)
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: OK, fixed.

File: doc/sphinx/practical-tools/coqide.rst (hunk 7 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst:53]: Yes. Is it still related to async mode? What should be the wording please?

File: doc/sphinx/practical-tools/coqide.rst (hunk 8 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst]: OK, fixed.

File: doc/sphinx/practical-tools/coqide.rst (hunk 10 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst:308]: All sections are covered, so I don’t understand your comment…?

File: doc/sphinx/practical-tools/coqide.rst (hunk 11 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst:346]: This single word alone is confusing for me…

File: doc/sphinx/practical-tools/coqide.rst (hunk 12 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst]: OK, fixed.

File: doc/sphinx/practical-tools/coqide.rst (hunk 13 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst]: Strange American usage. OK, restored.

File: doc/sphinx/practical-tools/coqide.rst (hunk 14 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst]: OK, restored.

**Comment by Frigory33** [comment]
> One other thing: we use several terms for "key bindings", including "Shortcuts" and "menu item accelerators". Lets at least drop "accelerator", which is used in only 2 places in the doc.

I wasn’t bothered by the use of synonyms, but OK, I’ve replaced “accelerators” with “shortcuts” in the panel.

Thanks a lot for your review!

File: doc/sphinx/practical-tools/coqide.rst (hunk 7 above)
Comment (jfehrle) [reply doc/sphinx/practical-tools/coqide.rst:53]: How about "- the *Warnings panel* shows warnings detected when running in :ref:`async mode <asyncmode>`;"

File: doc/sphinx/practical-tools/coqide.rst (hunk 10 above)
Comment (jfehrle) [reply doc/sphinx/practical-tools/coqide.rst:308]: By "cover", I meant to briefly document what Colors, Tags, Shortcuts and Misc are for.  They are not described here.  Only the ones that are checked off here are described:

image

File: doc/changelog/10-coqide/19417-coqide-preferences.rst (hunk 2 above)
Comment (Frigory33) [reply doc/changelog/10-coqide/19417-coqide-preferences.rst:22]: If it is really a problem of reusing the same word, I think the best thing to do is to use another word in “Editor appearance”, e.g. “Editor display”, “Editor view”, “Editor layout”.

File: ide/coqide/preferences.ml (hunk 16 above)
[the remark this answers is not in the document]
Comment (jfehrle) [reply ide/coqide/preferences.ml:85]: IMO better to allow for small variations as long as they are easy to read.  Reformatting tends to obscure changes in function.  Everyone has slightly different formatting preferences, so it's the appearance is not going to be perfectly consistent.  From what I've seen, the code base is consistent in indenting the cases rather than putting them at the left margin.

File: doc/sphinx/practical-tools/coqide.rst (hunk 10 above)
Comment (Frigory33) [reply doc/sphinx/practical-tools/coqide.rst:308]: To me, they are documented enough…

File: ide/coqide/preferences_ui.ml [hunk 29]
```diff
@@ -0,0 +1,690 @@
...
+      { label = "Text font"; box; apply = set_font }
+    )
+  in
+
+  let config_highlight =
+    let source_param text ids elem_get pref =
+      let id_to_name id =
+        match elem_get id with
+        | Some elem -> elem#name
+        | None -> ""
```
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences_ui.ml]: OK, so I’ve put `id` instead (it is what was shown before).

File: ide/coqide/preferences.ml (hunk 16 above)
[the remark this answers is not in the document]
Comment (Frigory33) [reply ide/coqide/preferences.ml:85]: OK, I’ve put the indentation. But let’s keep `function`, it’s the keyword to use here.

**Review by jfehrle — APPROVED** [review]
Thanks for the PR.  This looks better than what we had before!

@coqbot: merge now
